src/ops/telex.c

```diff
@@ -0,0 +1,654 @@
+const tele_op_t op_TI_PARAM_QT        = MAKE_GET_OP(TI.PARAM.QT         , op_TI_PARAM_QT_get        , 1, true);
+const tele_op_t op_TI_PARAM_N         = MAKE_GET_OP(TI.PARAM.N          , op_TI_PARAM_N_get         , 1, true);
+const tele_op_t op_TI_PARAM_SCALE     = MAKE_GET_OP(TI.PARAM.SCALE      , op_TI_PARAM_SCALE_get     , 2, false);
+const tele_op_t op_TI_PARAM_MAP       = MAKE_GET_OP(TI.PARAM.MAP        , op_TI_PARAM_MAP_get       , 3, false);
+
+const tele_op_t op_TI_IN              = MAKE_GET_OP(TI.IN               , op_TI_IN_get              , 1, true);
+const tele_op_t op_TI_IN_QT           = MAKE_GET_OP(TI.IN.QT            , op_TI_IN_QT_get           , 1, true);
+const tele_op_t op_TI_IN_N            = MAKE_GET_OP(TI.IN.N             , op_TI_IN_N_get            , 1, true);
+const tele_op_t op_TI_IN_SCALE        = MAKE_GET_OP(TI.IN.SCALE         , op_TI_IN_SCALE_get        , 2, false);
+const tele_op_t op_TI_IN_MAP          = MAKE_GET_OP(TI.IN.MAP           , op_TI_IN_MAP_get          , 3, false);
+
+const tele_op_t op_TI_PARAM_CALIB     = MAKE_GET_OP(TI.PARAM.CALIB      , op_TI_PARAM_CALIB_get     , 2, false);
+const tele_op_t op_TI_IN_CALIB        = MAKE_GET_OP(TI.IN.CALIB         , op_TI_IN_CALIB_get        , 2, false);
+const tele_op_t op_TI_STORE           = MAKE_GET_OP(TI.STORE            , op_TI_STORE_get           , 1, false);
+const tele_op_t op_TI_RESET           = MAKE_GET_OP(TI.RESET            , op_TI_RESET_get           , 1, false);
+
+
+// clang-format on
+
+// telex helpers
+void TXSend(uint8_t model, uint8_t command, uint8_t output, int16_t value,
+            bool set) {
+    // zero-index the output
+    output -= 1;
+    // convert the output to the device and the port
+    uint8_t port = output % 4;
+    uint8_t device = output >> 2;
+    uint8_t address = model + device;
+    // init and fill the buffer	(make the buffer smaller if we are not sending a
+    // payload)
+    uint8_t buffer[set ? 4 : 2];
+    buffer[0] = command;
+    buffer[1] = port;
+    if (set) {
+        uint16_t temp = (uint16_t)value;
+        buffer[2] = temp >> 8;
+        buffer[3] = temp & 0xff;
+    }
+    tele_ii_tx(address, buffer, set ? 4 : 2);
+}
+void TXCmd(uint8_t model, uint8_t command, uint8_t output) {
+    TXSend(model, command, output, 0, false);
+}
+void TXSet(uint8_t model, uint8_t command, command_state_t *cs) {
+    uint8_t output = cs_pop(cs);
+    int16_t value = cs_pop(cs);
+    TXSend(model, command, output, value, true);
+}
+void TXReceive(uint8_t model, command_state_t *cs, uint8_t mode, bool shift) {
+    // zero-index the output
+    uint8_t input = cs_pop(cs) - 1;
+    // send the port, device and address
+    uint8_t port = input % 4;
+    uint8_t device = input >> 2;
+    uint8_t address = model + device;
+    // inputs are numbered 0-7 for each device - shift is for the second half
+    // mode pushes it up so it can read quantized values and note numbers
+    port += (shift ? 4 : 0) + (mode * 8);
```

Comment:
Another great find! Thank you so much for taking the time to put your eyes to this. Sincerely appreciated.